flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryRow.java

```diff
@@ -63,6 +91,22 @@ public static int calculateFixPartSizeInBytes(int arity) {
 		return calculateBitSetWidthInBytes(arity) + 8 * arity;
 	}
 
+	/**
+	 * If it is a fixed-length field, we can call this BinaryRow's setXX method for in-place updates.
+	 * If it is variable-length field, can't use this method, because the underlying data is stored continuously.
+	 */
+	public static boolean isFixedLength(InternalType type) {
+		if (type instanceof DecimalType) {
+			return ((DecimalType) type).precision() <= DecimalType.MAX_COMPACT_PRECISION;
+		} else {
+			return MUTABLE_FIELD_TYPES.contains(type);
+		}
+	}
+
+	public static boolean isMutable(InternalType type) {
+		return MUTABLE_FIELD_TYPES.contains(type) || type instanceof DecimalType;
```

Comment:
do we also need to check Decimal's precision just like `isFixedLength` does?

Reply:
No matter how precision is, it is mutable.
Precision is compact, Decimal is stored in 8 bytes..
Precision is not compact, Decimal is stored in 16 bytes.

Reply:
method isFixedLength should be isInFixedLengthPart

Reply:
ok,

Reply:
why not just add the dicimal type to the MUTABLE_FIELD_TYPES, and give some explainations

Reply:
Dicimal type can not be added into MUTABLE_FIELD_TYPES. Because its instance is not unique, but it has different instances for different `precision` or `scale` value.